packages/tenant-process/src/services/readModelService.ts

```diff
@@ -33,4 +33,108 @@ async function getTotalCount(
 }
 */
 
-export const readModelService = {};
+export const readModelService = {
+  async getTenant(id: string): Promise<WithMetadata<Tenant> | undefined> {
+    const data = await tenants.findOne(
+      { "data.id": id },
+      { projection: { data: true, metadata: true } }
+    );
+    if (data) {
+      const result = z
+        .object({
+          metadata: z.object({ version: z.number() }),
+          data: Tenant,
+        })
+        .safeParse(data);
+
+      if (!result.success) {
+        logger.error(
+          `Unable to parse tenant item: result ${JSON.stringify(
+            result
+          )} - data ${JSON.stringify(data)} `
+        );
+
+        throw ErrorTypes.GenericError;
+      }
+
+      return {
+        data: result.data.data,
+        metadata: { version: result.data.metadata.version },
+      };
+    }
+
+    return undefined;
+  },
+
+  async getTenantByExternalId({
+    origin,
+    code,
+  }: {
+    origin: string;
+    code: string;
+  }): Promise<WithMetadata<Tenant> | undefined> {
+    const data = await tenants.findOne(
+      {
+        "data.externalId.value": code,
+        "data.externalId.origin": origin,
+      },
+      { projection: { data: true, metadata: true } }
+    );
+    if (data) {
+      const result = z
+        .object({
+          metadata: z.object({ version: z.number() }),
+          data: Tenant,
+        })
+        .safeParse(data);
+      if (!result.success) {
+        logger.error(
+          `Unable to parse tenant item: result ${JSON.stringify(
+            result
+          )} - data ${JSON.stringify(data)} `
+        );
+        throw ErrorTypes.GenericError;
+      }
+      return {
+        data: result.data.data,
+        metadata: { version: result.data.metadata.version },
+      };
+    }
+    return undefined;
+  },
+
+  async getTenantBySelfcareId(
+    selfcareId: string
+  ): Promise<WithMetadata<Tenant> | undefined> {
+    const data = await tenants.findOne(
+      { "data.selfcareId": selfcareId },
+      { projection: { data: true, metadata: true } }
+    );
+
+    if (data) {
+      const result = z
+        .object({
+          metadata: z.object({ version: z.number() }),
+          data: Tenant,
+        })
+        .safeParse(data);
+
+      if (!result.success) {
+        logger.error(
+          `Unable to parse tenant item: result ${JSON.stringify(
+            result
+          )} - data ${JSON.stringify(data)} `
+        );
+
+        throw ErrorTypes.GenericError;
+      }
+
+      return {
+        data: result.data.data,
+        metadata: { version: result.data.metadata.version },
+      };
+    }
+
+    return undefined;
+  },
+};
```

Comment:
I prefer avoiding code duplication
(I wrote it without an IDE, please check if it actually works 😄 )
```suggestion

async function getTenant(filter: NotSureAboutTheTypeHere): Promise<WithMetadata<Tenant> | undefined> {
    const data = await tenants.findOne(
      filter,
      { projection: { data: true, metadata: true } }
    );
    if (data) {
      const result = z
        .object({
          metadata: z.object({ version: z.number() }),
          data: Tenant,
        })
        .safeParse(data);

      if (!result.success) {
        logger.error(
          `Unable to parse tenant item: result ${JSON.stringify(
            result
          )} - data ${JSON.stringify(data)} `
        );

        throw ErrorTypes.GenericError;
      }

      return {
        data: result.data.data,
        metadata: { version: result.data.metadata.version },
      };
    }

    return undefined;
  }

export const readModelService = {
  async getTenantById(id: string): Promise<WithMetadata<Tenant> | undefined> {
    return getTenant({ "data.id": id })
  },

  async getTenantByExternalId(origin: string, code: string): Promise<WithMetadata<Tenant> | undefined> {
    return getTenant( {
        "data.externalId.value": code,
        "data.externalId.origin": origin,
      })
  },

  async getTenantBySelfcareId(
    selfcareId: string
  ): Promise<WithMetadata<Tenant> | undefined> {
    return getTenant({ "data.selfcareId": selfcareId })
  },
};
```

Reply:
I had already tried to eliminate the repeating code, but I had difficulty with the types,
I solved it by doing this: 

```
type TenantInput =
  | { "data.id": string }
  | {
      "data.externalId.value": string;
      "data.externalId.origin": string;
    }
  | { "data.selfcareId": string };
```

It's not very elegant but it works.
I hope you like it.
I did the refactor in the last commit.

Reply:
IMO the limit of this solution is the increased portion of code to touch for each change, without the additional benefits.

I suggest to:
- in the `getTenant` signature leave the input type without constraints (you can remove `TenantInput` and use `Filter<{data: Tenant}>` for the input type)
- as per my example, do not export the `getTenant` function. This allows to keep control on the queries that are executed with this service, without limiting the filters

Reply:
Ok, thanks for the suggest.
I did as you say

Reply:
I created a type util that could help us in this case

<img width="934" alt="Screenshot 2023-11-08 alle 09 51 14" src="https://github.com/pagopa/interop-be-monorepo/assets/62668966/2e045ffb-5a34-44b9-90e6-8bfcd8cb8a9e">

Reply:
🤩